content/en/blog/_posts/advanced-server-side-apply.md

```diff
@@ -0,0 +1,80 @@
+
+If your controller is perfect, great! You don’t need to change it. But if you do want to change it, you can take advantage of the new client library’s “extract” workflow– that is, get the existing object, extract your existing desires, make modifications, and re-apply. For many controllers that were computing the smallest changes possible, this will be a small change.
+
+This workflow avoids the failure mode of accidentally trying to own every field in the object, which is what happens if you just GET the object, make changes, and then APPLY. (Note that the server will notice you did this and reject your change!)
+
+#### Reconstructive controllers
+
+This kind of controller wasn’t really possible prior to SSA. The idea here is to (whenever something changes etc) reconstruct from scratch the fields of the object as the controller wishes them to be, and then APPLY the change to the server, letting it figure out the result. I now recommend that new controllers start out this way–it’s less fiddly to say what you want an object to look like than it is to say how you want it to change.
+
+The client library supports this method of operation by default.
+
+The only downside is that you may end up sending unneeded APPLY requests to apiserver, if actually the object already matches your controller’s desires. This doesn’t matter if it happens once in a while, but for extremely high-throughput controllers, it might cause a performance problem for the cluster–specifically, apiserver. No-op writes are not written to storage (etcd) or broadcast to any watchers, so it’s not really that big of a deal. If you’re worried about this anyway, today you could use the method explained in the previous section, or you could still do it this way for now, and wait for an additional client-side mechanism to suppress zero-change suppress.
+
+To get around this downside, why not GET the object and only send your APPLY if the object needs it? Surprisingly, it doesn’t help much– a no-op APPLY is not very much more work for apiserver than an extra GET; and an APPLY that changes things is cheaper than that same APPLY with a preceding GET. Worse, since it is a distributed system, something could change between your GET and APPLY, invalidating your computation. Instead, we can use this optimization on an object retrieved from a cache–then it legitimately will reduce load on the system (at the cost of a delay when a change is needed and the cache is a bit behind).
+
+#### CI/CD systems
+Continuous integration and/or deployment systems are a special kind of controller which is doing something like reading source .yaml files from a git repo and automatically pushing them into the cluster, perhaps after running them through some testing, or waiting for test results, etc. Generally users are the entities pushing changes into the git repo, although that’s not necessarily always the case.
```

Comment:
```suggestion
Continuous integration (CI) and/or continuous deployment (CD) systems are a special kind of controller which is doing something like reading manifests from a Git repo and automatically pushing them into the cluster. Perhaps the CI / CD process first generates manifests from a template, then runs some tests, and then deploys a change. Typically, users are the entities pushing changes into the Git repository, although that’s not necessarily always the case.
```

Reply:
Included most of this but also applied the "source control" from another comment.